{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dashboard.tabs.riboseq_atlas import get_average_coverage\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/root/protein_tools_input_ensembl_proteins_alignments.m8 exists and will be overwritten\n",
      "easy-search --format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits /root/protein_tools_input.fasta /root/ensembl_proteins.fa /root/protein_tools_input_ensembl_proteins_alignments.m8 /root/tmp \n",
      "\n",
      "MMseqs Version:                        \tad5837b3444728411e6c90f8c6ba9370f665c443\n",
      "Substitution matrix                    \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t3\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \t1\n",
      "Pseudo count b                         \t1.5\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Gap open cost                          \tnucl:5,aa:11\n",
      "Gap extension cost                     \tnucl:2,aa:1\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t16\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                            \t5.7\n",
      "k-mer length                           \t0\n",
      "k-score                                \t2147483647\n",
      "Alphabet size                          \tnucl:5,aa:21\n",
      "Max results per query                  \t300\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.001\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \ttrue\n",
      "Alignment format                       \t0\n",
      "Format alignment output                \tquery,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits\n",
      "Database output                        \tfalse\n",
      "Overlap threshold                      \t0\n",
      "Database type                          \t0\n",
      "Shuffle input database                 \ttrue\n",
      "Createdb mode                          \t0\n",
      "Write lookup file                      \t0\n",
      "Greedy best hits                       \tfalse\n",
      "\n",
      "createdb /root/protein_tools_input.fasta /root/tmp/10458996324427652754/query --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to query_h: 0h 0m 0s 2ms\n",
      "Time for merging to query: 0h 0m 0s 2ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "createdb /root/ensembl_proteins.fa /root/tmp/10458996324427652754/target --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[=======\n",
      "Time for merging to target_h: 0h 0m 0s 8ms\n",
      "Time for merging to target: 0h 0m 0s 26ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 181ms\n",
      "Create directory /root/tmp/10458996324427652754/search_tmp\n",
      "search /root/tmp/10458996324427652754/query /root/tmp/10458996324427652754/target /root/tmp/10458996324427652754/result /root/tmp/10458996324427652754/search_tmp --alignment-mode 3 -s 5.7 --remove-tmp-files 1 \n",
      "\n",
      "prefilter /root/tmp/10458996324427652754/query /root/tmp/10458996324427652754/target /root/tmp/10458996324427652754/search_tmp/11863743554009048031/pref_0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 300 --split 0 --split-mode 2 --split-memory-limit 0 -c 0 --cov-mode 0 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 16 --compressed 0 -v 3 -s 5.7 \n",
      "\n",
      "Query database size: 5606 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 77374 type: Aminoacid\n",
      "Index table k-mer threshold: 112 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 77.37K 0s 775ms\n",
      "Index table: Masked residues: 1228104\n",
      "Index table: fill\n",
      "[=================================================================] 77.37K 0s 550ms\n",
      "Index statistics\n",
      "Entries:          25256013\n",
      "DB size:          632 MB\n",
      "Avg k-mer size:   0.394625\n",
      "Top 10 k-mers\n",
      "    DVVFEE\t434\n",
      "    VMEYLV\t433\n",
      "    LYDMNY\t404\n",
      "    TFDAFS\t366\n",
      "    YRVLYR\t256\n",
      "    VAESEW\t236\n",
      "    PMFLLS\t210\n",
      "    GEVLSS\t198\n",
      "    VTSSSS\t197\n",
      "    TFDAFT\t192\n",
      "Time for index table init: 0h 0m 1s 920ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 112\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 5606\n",
      "Target db start 1 to 77374\n",
      "[=================================================================] 5.61K 0s 403ms\n",
      "\n",
      "274.012665 k-mers per position\n",
      "5551 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "55 sequences passed prefiltering per query sequence\n",
      "35 median result list length\n",
      "294 sequences with 0 size result lists\n",
      "Time for merging to pref_0: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 2s 733ms\n",
      "align /root/tmp/10458996324427652754/query /root/tmp/10458996324427652754/target /root/tmp/10458996324427652754/search_tmp/11863743554009048031/pref_0 /root/tmp/10458996324427652754/result --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 5606 type: Aminoacid\n",
      "Target database size: 77374 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 5.61K 0s 804ms\n",
      "Time for merging to result: 0h 0m 0s 2ms\n",
      "311188 alignments calculated\n",
      "22082 sequence pairs passed the thresholds (0.070960 of overall calculated)\n",
      "3.938994 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 821ms\n",
      "rmdb /root/tmp/10458996324427652754/search_tmp/11863743554009048031/pref_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/10458996324427652754/search_tmp/11863743554009048031/aln_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/10458996324427652754/search_tmp/11863743554009048031/input_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/10458996324427652754/search_tmp/11863743554009048031/aln_merge -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "/root/protein_tools_input_ensembl_proteins_alignments.m8 exists and will be overwritten\n",
      "convertalis /root/tmp/10458996324427652754/query /root/tmp/10458996324427652754/target /root/tmp/10458996324427652754/result /root/protein_tools_input_ensembl_proteins_alignments.m8 --sub-mat nucl:nucleotide.out,aa:blosum62.out --format-mode 0 --format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits --translation-table 1 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 5.61K 0s 11ms\n",
      "Time for merging to protein_tools_input_ensembl_proteins_alignments.m8: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 35ms\n",
      "rmdb /root/tmp/10458996324427652754/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/10458996324427652754/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb /root/tmp/10458996324427652754/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/10458996324427652754/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/10458996324427652754/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "/root/protein_tools_input_swissprot_proteins_alignments.m8 exists and will be overwritten\n",
      "easy-search --format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits /root/protein_tools_input.fasta /root/swissprot_proteins.fa /root/protein_tools_input_swissprot_proteins_alignments.m8 /root/tmp \n",
      "\n",
      "MMseqs Version:                        \tad5837b3444728411e6c90f8c6ba9370f665c443\n",
      "Substitution matrix                    \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t3\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \t1\n",
      "Pseudo count b                         \t1.5\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Gap open cost                          \tnucl:5,aa:11\n",
      "Gap extension cost                     \tnucl:2,aa:1\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t16\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                            \t5.7\n",
      "k-mer length                           \t0\n",
      "k-score                                \t2147483647\n",
      "Alphabet size                          \tnucl:5,aa:21\n",
      "Max results per query                  \t300\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.001\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \ttrue\n",
      "Alignment format                       \t0\n",
      "Format alignment output                \tquery,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits\n",
      "Database output                        \tfalse\n",
      "Overlap threshold                      \t0\n",
      "Database type                          \t0\n",
      "Shuffle input database                 \ttrue\n",
      "Createdb mode                          \t0\n",
      "Write lookup file                      \t0\n",
      "Greedy best hits                       \tfalse\n",
      "\n",
      "createdb /root/protein_tools_input.fasta /root/tmp/1160841380439698130/query --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to query_h: 0h 0m 0s 2ms\n",
      "Time for merging to query: 0h 0m 0s 2ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "createdb /root/swissprot_proteins.fa /root/tmp/1160841380439698130/target --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[==\n",
      "Time for merging to target_h: 0h 0m 0s 3ms\n",
      "Time for merging to target: 0h 0m 0s 10ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 66ms\n",
      "Create directory /root/tmp/1160841380439698130/search_tmp\n",
      "search /root/tmp/1160841380439698130/query /root/tmp/1160841380439698130/target /root/tmp/1160841380439698130/result /root/tmp/1160841380439698130/search_tmp --alignment-mode 3 -s 5.7 --remove-tmp-files 1 \n",
      "\n",
      "prefilter /root/tmp/1160841380439698130/query /root/tmp/1160841380439698130/target /root/tmp/1160841380439698130/search_tmp/18160661014718753080/pref_0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 300 --split 0 --split-mode 2 --split-memory-limit 0 -c 0 --cov-mode 0 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 16 --compressed 0 -v 3 -s 5.7 \n",
      "\n",
      "Query database size: 5606 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 20330 type: Aminoacid\n",
      "Index table k-mer threshold: 112 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 20.33K 0s 330ms\n",
      "Index table: Masked residues: 557826\n",
      "Index table: fill\n",
      "[=================================================================] 20.33K 0s 223ms\n",
      "Index statistics\n",
      "Entries:          10408624\n",
      "DB size:          547 MB\n",
      "Avg k-mer size:   0.162635\n",
      "Top 10 k-mers\n",
      "    DRVIPL\t212\n",
      "    PMFLLS\t194\n",
      "    KASCHL\t145\n",
      "    MLPILR\t140\n",
      "    LHPYLS\t130\n",
      "    VMEYLV\t130\n",
      "    LYDMNY\t123\n",
      "    RDMERN\t71\n",
      "    LAMYYV\t69\n",
      "    SSSSSS\t65\n",
      "Time for index table init: 0h 0m 1s 120ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 112\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 5606\n",
      "Target db start 1 to 20330\n",
      "[=================================================================] 5.61K 0s 369ms\n",
      "\n",
      "274.012665 k-mers per position\n",
      "2301 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "21 sequences passed prefiltering per query sequence\n",
      "14 median result list length\n",
      "389 sequences with 0 size result lists\n",
      "Time for merging to pref_0: 0h 0m 0s 11ms\n",
      "Time for processing: 0h 0m 1s 897ms\n",
      "align /root/tmp/1160841380439698130/query /root/tmp/1160841380439698130/target /root/tmp/1160841380439698130/search_tmp/18160661014718753080/pref_0 /root/tmp/1160841380439698130/result --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 5606 type: Aminoacid\n",
      "Target database size: 20330 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 5.61K 0s 390ms\n",
      "Time for merging to result: 0h 0m 0s 9ms\n",
      "119470 alignments calculated\n",
      "7680 sequence pairs passed the thresholds (0.064284 of overall calculated)\n",
      "1.369961 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 415ms\n",
      "rmdb /root/tmp/1160841380439698130/search_tmp/18160661014718753080/pref_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/1160841380439698130/search_tmp/18160661014718753080/aln_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/1160841380439698130/search_tmp/18160661014718753080/input_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/1160841380439698130/search_tmp/18160661014718753080/aln_merge -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "/root/protein_tools_input_swissprot_proteins_alignments.m8 exists and will be overwritten\n",
      "convertalis /root/tmp/1160841380439698130/query /root/tmp/1160841380439698130/target /root/tmp/1160841380439698130/result /root/protein_tools_input_swissprot_proteins_alignments.m8 --sub-mat nucl:nucleotide.out,aa:blosum62.out --format-mode 0 --format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits --translation-table 1 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 5.61K 0s 8ms\n",
      "Time for merging to protein_tools_input_swissprot_proteins_alignments.m8: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 23ms\n",
      "rmdb /root/tmp/1160841380439698130/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/1160841380439698130/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /root/tmp/1160841380439698130/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/1160841380439698130/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/1160841380439698130/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "/root/protein_tools_input_GRCh38_latest_protein_alignments.m8 exists and will be overwritten\n",
      "easy-search --format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits /root/protein_tools_input.fasta /root/GRCh38_latest_protein.faa /root/protein_tools_input_GRCh38_latest_protein_alignments.m8 /root/tmp \n",
      "\n",
      "MMseqs Version:                        \tad5837b3444728411e6c90f8c6ba9370f665c443\n",
      "Substitution matrix                    \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t3\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \t1\n",
      "Pseudo count b                         \t1.5\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Gap open cost                          \tnucl:5,aa:11\n",
      "Gap extension cost                     \tnucl:2,aa:1\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t16\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                            \t5.7\n",
      "k-mer length                           \t0\n",
      "k-score                                \t2147483647\n",
      "Alphabet size                          \tnucl:5,aa:21\n",
      "Max results per query                  \t300\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.001\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \ttrue\n",
      "Alignment format                       \t0\n",
      "Format alignment output                \tquery,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits\n",
      "Database output                        \tfalse\n",
      "Overlap threshold                      \t0\n",
      "Database type                          \t0\n",
      "Shuffle input database                 \ttrue\n",
      "Createdb mode                          \t0\n",
      "Write lookup file                      \t0\n",
      "Greedy best hits                       \tfalse\n",
      "\n",
      "createdb /root/protein_tools_input.fasta /root/tmp/8241599462884776445/query --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to query_h: 0h 0m 0s 2ms\n",
      "Time for merging to query: 0h 0m 0s 2ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "createdb /root/GRCh38_latest_protein.faa /root/tmp/8241599462884776445/target --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[=============\n",
      "Time for merging to target_h: 0h 0m 0s 19ms\n",
      "Time for merging to target: 0h 0m 0s 74ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 479ms\n",
      "Create directory /root/tmp/8241599462884776445/search_tmp\n",
      "search /root/tmp/8241599462884776445/query /root/tmp/8241599462884776445/target /root/tmp/8241599462884776445/result /root/tmp/8241599462884776445/search_tmp --alignment-mode 3 -s 5.7 --remove-tmp-files 1 \n",
      "\n",
      "prefilter /root/tmp/8241599462884776445/query /root/tmp/8241599462884776445/target /root/tmp/8241599462884776445/search_tmp/6878996669703474559/pref_0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 300 --split 0 --split-mode 2 --split-memory-limit 0 -c 0 --cov-mode 0 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 16 --compressed 0 -v 3 -s 5.7 \n",
      "\n",
      "Query database size: 5606 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 136193 type: Aminoacid\n",
      "Index table k-mer threshold: 112 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 136.19K 2s 693ms\n",
      "Index table: Masked residues: 4264446\n",
      "Index table: fill\n",
      "[=================================================================] 136.19K 1s 934ms\n",
      "Index statistics\n",
      "Entries:          87074974\n",
      "DB size:          986 MB\n",
      "Avg k-mer size:   1.360546\n",
      "Top 10 k-mers\n",
      "    LYDMNY\t1020\n",
      "    VMEYLV\t820\n",
      "    LEYNSL\t708\n",
      "    SSPSSS\t603\n",
      "    SAKPKV\t511\n",
      "    LDARYR\t495\n",
      "    SSSLSS\t464\n",
      "    GLSSLT\t462\n",
      "    VHSGRT\t450\n",
      "    PARLDV\t433\n",
      "Time for index table init: 0h 0m 5s 320ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 112\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 5606\n",
      "Target db start 1 to 136193\n",
      "[=================================================================] 5.61K 0s 531ms\n",
      "\n",
      "274.012665 k-mers per position\n",
      "18648 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "129 sequences passed prefiltering per query sequence\n",
      "102 median result list length\n",
      "250 sequences with 0 size result lists\n",
      "Time for merging to pref_0: 0h 0m 0s 5ms\n",
      "Time for processing: 0h 0m 6s 303ms\n",
      "align /root/tmp/8241599462884776445/query /root/tmp/8241599462884776445/target /root/tmp/8241599462884776445/search_tmp/6878996669703474559/pref_0 /root/tmp/8241599462884776445/result --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 5606 type: Aminoacid\n",
      "Target database size: 136193 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 5.61K 2s 422ms\n",
      "Time for merging to result: 0h 0m 0s 9ms\n",
      "725393 alignments calculated\n",
      "31364 sequence pairs passed the thresholds (0.043237 of overall calculated)\n",
      "5.594720 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 461ms\n",
      "rmdb /root/tmp/8241599462884776445/search_tmp/6878996669703474559/pref_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /root/tmp/8241599462884776445/search_tmp/6878996669703474559/aln_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/8241599462884776445/search_tmp/6878996669703474559/input_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/8241599462884776445/search_tmp/6878996669703474559/aln_merge -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "/root/protein_tools_input_GRCh38_latest_protein_alignments.m8 exists and will be overwritten\n",
      "convertalis /root/tmp/8241599462884776445/query /root/tmp/8241599462884776445/target /root/tmp/8241599462884776445/result /root/protein_tools_input_GRCh38_latest_protein_alignments.m8 --sub-mat nucl:nucleotide.out,aa:blosum62.out --format-mode 0 --format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits --translation-table 1 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 5.61K 0s 17ms\n",
      "Time for merging to protein_tools_input_GRCh38_latest_protein_alignments.m8: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 58ms\n",
      "rmdb /root/tmp/8241599462884776445/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/8241599462884776445/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb /root/tmp/8241599462884776445/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /root/tmp/8241599462884776445/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /root/tmp/8241599462884776445/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [tblastn] Query_2409 VTX-0850133: Could not calculate ungapped Karlin-Altschul parameters due to an invalid query sequence or its translation. Please verify the query sequence(s) and/or filtering options \n",
      "Command line argument error: Argument \"query\". File is not accessible:  `/data/protein_tools_input'\n",
      "Warning: [tblastn] Query_434 VTX-0851616: Could not calculate ungapped Karlin-Altschul parameters due to an invalid query sequence or its translation. Please verify the query sequence(s) and/or filtering options \n",
      "Warning: [tblastn] Query_1505 VTX-0850135: Could not calculate ungapped Karlin-Altschul parameters due to an invalid query sequence or its translation. Please verify the query sequence(s) and/or filtering options \n"
     ]
    }
   ],
   "source": [
    "#     from dashboard import plotting, description\n",
    "#     from dashboard.etl.sorf_query import load_jsonlines_table\n",
    "    \n",
    "#     import gzip\n",
    "#     import shlex\n",
    "    \n",
    "#     import pathlib\n",
    "#     import subprocess\n",
    "#     import pandas as pd\n",
    "#     import click\n",
    "    \n",
    "#     input_vtx = pathlib.Path('/home/ec2-user/repos/dashboard/cache/protein_data/protein_tools_input.fasta')\n",
    "#     output_prefix = pathlib.Path('/home/ec2-user/repos/dashboard/cache/protein_data')\n",
    "\n",
    "# # input_vtx = #pathlib.Path('/home/ec2-user/repos/dashboard/cache/protein_data/protein_tools_input.fasta')\n",
    "# # CACHE_DIR = pathlib.Path('/home/ec2-user/repos/dashboard/cache')\n",
    "# # output_prefix = pathlib.Path('/home/ec2-user/repos/dashboard/cache/protein_data')\n",
    "\n",
    "def run_protein_search_tools(input_vtx, output_prefix):\n",
    "    input_vtx = pathlib.Path(input_vtx)\n",
    "    output_prefix = pathlib.Path(output_prefix)\n",
    "    \n",
    "    REFSEQ_FASTA_FILE = output_prefix.joinpath('GRCh38_latest_protein.faa')\n",
    "    ENSEMBL_FASTA_FILE = output_prefix.joinpath('swissprot_proteins.fa')\n",
    "    SWISSPROT_FASTA_FILE = output_prefix.joinpath('ensembl_proteins.fa')\n",
    "    VTX_WITHOUT_SIGNAL_SEQ = output_prefix.joinpath('vtx_aa_seq_signal_sequence_removed.fa')\n",
    "    any_missing = []\n",
    "    for f in [REFSEQ_FASTA_FILE, ENSEMBL_FASTA_FILE, SWISSPROT_FASTA_FILE, VTX_WITHOUT_SIGNAL_SEQ]:\n",
    "        if not f.exists():\n",
    "            any_missing.append(f)\n",
    "    if len(any_missing) > 0:\n",
    "        raise ValueError(f\"Reference files are missing the following files don't exist. \\\n",
    "            please check that update_cache.py has completed, or create these manually in the output_prefix directory.: {any_missing}\")\n",
    "    # Swissprot Isoform Check\n",
    "    fasta_file = SWISSPROT_FASTA_FILE\n",
    "    query_db = f'/root/{input_vtx.name}'\n",
    "    target_db = f'/root/{fasta_file.name}'\n",
    "    output_file = f'/root/{input_vtx.stem}_{fasta_file.stem}_alignments.m8'\n",
    "    options = '--format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits'\n",
    "\n",
    "    base_cmd = f\"docker run --rm -v {output_prefix}:/root soedinglab/mmseqs2 mmseqs easy-search\" \n",
    "    full_cmd = f'{base_cmd} {options} {query_db} {target_db} {output_file} /root/tmp'\n",
    "    subprocess.run(shlex.split(full_cmd))\n",
    "\n",
    "    isoform_df = pd.read_csv(output_prefix.joinpath(f'{input_vtx.stem}_{fasta_file.stem}_alignments.m8'), sep='\\t', names=options.split()[1].split(','))\n",
    "    isoform_df = isoform_df[(isoform_df['pident'] == 100.0)]\n",
    "    isoform_df.drop_duplicates(inplace=True)\n",
    "    swissprot_isoform_df = isoform_df.groupby('query').aggregate(list)\n",
    "    swissprot_isoform_df.rename(columns={'target': 'swissprot_isoform'}, inplace=True)\n",
    "\n",
    "    # Ensembl Isoforms\n",
    "    fasta_file = ENSEMBL_FASTA_FILE\n",
    "    query_db = f'/root/{input_vtx.name}'\n",
    "    target_db = f'/root/{fasta_file.name}'\n",
    "    output_file = f'/root/{input_vtx.stem}_{fasta_file.stem}_alignments.m8'\n",
    "    options = '--format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits'\n",
    "\n",
    "    base_cmd = f\"docker run --rm -v {output_prefix}:/root soedinglab/mmseqs2 mmseqs easy-search\" \n",
    "    full_cmd = f'{base_cmd} {options} {query_db} {target_db} {output_file} /root/tmp'\n",
    "    subprocess.run(shlex.split(full_cmd))\n",
    "    \n",
    "    isoform_df = pd.read_csv(output_prefix.joinpath(f'{input_vtx.stem}_{fasta_file.stem}_alignments.m8'), sep='\\t', names=options.split()[1].split(','))\n",
    "    isoform_df = isoform_df[(isoform_df['pident'] == 100.0)]\n",
    "    isoform_df.drop_duplicates(inplace=True)\n",
    "    ensembl_isoform_df = isoform_df.groupby('query').aggregate(list)\n",
    "    ensembl_isoform_df.rename(columns={'target': 'ensembl_isoform'}, inplace=True)\n",
    "\n",
    "    # Refseq Isoforms\n",
    "    fasta_file = REFSEQ_FASTA_FILE\n",
    "    query_db = f'/root/{input_vtx.name}'\n",
    "    target_db = f'/root/{fasta_file.name}'\n",
    "    output_file = f'/root/{input_vtx.stem}_{fasta_file.stem}_alignments.m8'\n",
    "    options = '--format-output query,target,evalue,qstart,qend,qlen,qcov,gapopen,pident,fident,alnlen,raw,bits'\n",
    "\n",
    "    base_cmd = f\"docker run --rm -v {output_prefix}:/root soedinglab/mmseqs2 mmseqs easy-search\" \n",
    "    full_cmd = f'{base_cmd} {options} {query_db} {target_db} {output_file} /root/tmp'\n",
    "    subprocess.run(shlex.split(full_cmd))\n",
    "    \n",
    "    isoform_df = pd.read_csv(output_prefix.joinpath(f'{input_vtx.stem}_{fasta_file.stem}_alignments.m8'), sep='\\t', names=options.split()[1].split(','))\n",
    "    isoform_df = isoform_df[(isoform_df['pident'] == 100.0)]\n",
    "    isoform_df.drop_duplicates(inplace=True)\n",
    "\n",
    "    refseq_isoform_df = isoform_df.groupby('query').aggregate(list)\n",
    "    refseq_isoform_df.rename(columns={'target': 'refseq_isoform'}, inplace=True)\n",
    "\n",
    "#     swissprot_isoform_df.to_csv(pathlib.Path(output_prefix) / 'swissprot_isoform.csv')\n",
    "#     ensembl_isoform_df.to_csv(pathlib.Path(output_prefix) / 'ensembl_isoform.csv')\n",
    "#     refseq_isoform_df.to_csv(pathlib.Path(output_prefix) / 'refseq_isoform.csv')\n",
    "    \n",
    "    # Blast Tools\n",
    "    header = ['vtx_id', 'blastp_hit_id', 'blastp_description', 'blastp_score',\n",
    "          'blastp_query_coverage', 'blastp_align_length', 'blastp_align_identity', \n",
    "          'blastp_gaps', 'blastp_evalue']\n",
    "    blast_db_path = pathlib.Path('/efs/databases/blast')\n",
    "    blast_db = '-db mouse.protein.genbank.faa'\n",
    "    output_fmt = '6 qaccver saccver stitle bitscore qcovs length pident gaps evalue'\n",
    "    options = f'-outfmt \"{output_fmt}\" -num_threads 8'\n",
    "    query = f'-query /blast/data/{input_vtx.name}'\n",
    "    output = f'-out /blast/data/{input_vtx.stem}.blastp.out'\n",
    "    base_cmd = f'docker run --rm -v {blast_db_path}:/blast/blastdb -v {output_prefix}:/blast/data ncbi/blast'\n",
    "    full_cmd = f'{base_cmd} blastp {options} {blast_db} {query} {output}'\n",
    "    subprocess.run(shlex.split(full_cmd))\n",
    "    \n",
    "    blastp_df = pd.read_csv(output_prefix.joinpath(f'{input_vtx.stem}.blastp.out'), sep='\\t', names=header)\n",
    "    bdf = blastp_df.sort_values(by='blastp_score', ascending=False).groupby('vtx_id').first()\n",
    "    bdf.to_parquet(pathlib.Path(output_prefix).joinpath('mouse_blastp.parq'))\n",
    "\n",
    "    header = ['vtx_id', 'tblastn_hit_id', 'tblastn_description', 'tblastn_score',\n",
    "          'tblastn_query_coverage', 'tblastn_align_length', 'tblastn_align_identity', \n",
    "          'tblastn_gaps', 'tblastn_evalue']\n",
    "    blast_db_path = pathlib.Path('/efs/databases/blast')\n",
    "    blast_db = '-db mouse.rna.fna'\n",
    "    output_fmt = '6 qaccver saccver stitle score qcovs length pident gaps evalue'\n",
    "    options = f'-outfmt \"{output_fmt}\" -num_threads 8'\n",
    "    query = f'-query /blast/data/{input_vtx.name}'\n",
    "    output = f'-out /blast/data/{input_vtx.stem}.tblastn.out'\n",
    "    \n",
    "    base_cmd = f'docker run --rm -v {blast_db_path}:/blast/blastdb -v {output_prefix}:/blast/data ncbi/blast'\n",
    "    full_cmd = f'{base_cmd} tblastn {options} {blast_db} {query} {output}'\n",
    "    subprocess.run(shlex.split(full_cmd))\n",
    "    \n",
    "\n",
    "    tblastn_df = pd.read_csv(output_prefix.joinpath(f'{input_vtx.stem}.tblastn.out'), sep='\\t', names=header)\n",
    "    tdf = tblastn_df.sort_values(by='tblastn_score', ascending=False).groupby('vtx_id').first()\n",
    "    tdf.to_parquet(pathlib.Path(output_prefix).joinpath('mouse_tblastn.parq'))\n",
    "    \n",
    "    subprocess.run(shlex.split(f'docker run  --rm -v {output_prefix}:/data -v /efs/databases/blast:/db ncbi/blast blastp -task blastp-fast -outfmt 15 -db /db/mouse.protein.faa -query /data/{input_vtx.stem} -max_target_seqs 20 -num_threads 8 -out /data/blastp.results.json'))\n",
    "    \n",
    "    # Merge isoform results\n",
    "    isoform_df = pd.DataFrame([swissprot_isoform_df['swissprot_isoform'], ensembl_isoform_df['ensembl_isoform'], refseq_isoform_df['refseq_isoform']]).T\n",
    "    # isoform_df.replace(pd.NA, 'None', inplace=True)\n",
    "    isoform_df.to_parquet(pathlib.Path(output_prefix).joinpath('isoforms_search.parq'))\n",
    "    \n",
    "    # Run blast against sequneces with signal peptide removed\n",
    "    #Blast Tools\n",
    "    blast_db_path = pathlib.Path('/efs/databases/blast')\n",
    "    blast_db = '-db mouse.protein.genbank.faa'\n",
    "    output_fmt = '6 qaccver saccver stitle bitscore qcovs length pident gaps evalue'\n",
    "    options = f'-outfmt \"{output_fmt}\" -num_threads 8'\n",
    "    query = f'-query /blast/data/{VTX_WITHOUT_SIGNAL_SEQ.name}'\n",
    "    output = f'-out /blast/data/{VTX_WITHOUT_SIGNAL_SEQ.stem}.blastp.out'\n",
    "    base_cmd = f'docker run --rm -v {blast_db_path}:/blast/blastdb -v {output_prefix}:/blast/data ncbi/blast'\n",
    "    full_cmd = f'{base_cmd} blastp {options} {blast_db} {query} {output}'\n",
    "    subprocess.run(shlex.split(full_cmd))\n",
    "    \n",
    "    header = [\n",
    "        'vtx_id', 'nonsig_blastp_refseq_id', 'nonsig_blastp_hit_id', 'nonsig_blastp_description', 'nonsig_blastp_score',\n",
    "        'nonsig_blastp_align_length', 'nonsig_blastp_align_identity', 'nonsig_blastp_gaps', 'nonsig_blastp_evalue']\n",
    "    \n",
    "    non_sig_blastp_df = pd.read_csv(output_prefix.joinpath(f'{VTX_WITHOUT_SIGNAL_SEQ.stem}.blastp.out'), sep='\\t', names=header)\n",
    "    \n",
    "    header = [\n",
    "        'vtx_id', 'nonsig_tblastn_refseq_id', 'nonsig_tblastn_hit_id', 'nonsig_tblastn_description', 'nonsig_tblastn_score',\n",
    "        'nonsig_tblastn_align_length', 'nonsig_tblastn_align_identity', 'nonsig_tblastn_gaps', 'nonsig_tblastn_evalue']\n",
    "    output = f'-out /blast/data/{VTX_WITHOUT_SIGNAL_SEQ.stem}.tblastn.out'\n",
    "    blast_db = '-db mouse.rna.fna'\n",
    "    base_cmd = f'docker run --rm -v {blast_db_path}:/blast/blastdb -v {output_prefix}:/blast/data ncbi/blast'\n",
    "    full_cmd = f'{base_cmd} tblastn {options} {blast_db} {query} {output}'\n",
    "    subprocess.run(shlex.split(full_cmd))\n",
    "    tblastn_df = pd.read_csv(output_prefix.joinpath(f'{VTX_WITHOUT_SIGNAL_SEQ.stem}.tblastn.out'), sep='\\t', names=header)\n",
    "    tdf = tblastn_df.sort_values(by='nonsig_tblastn_score', ascending=False).groupby('vtx_id').first()\n",
    "    bdf = non_sig_blastp_df.sort_values('nonsig_blastp_score', ascending=False).groupby('vtx_id').first()\n",
    "    output_nonsig_df = bdf.merge(tdf, left_index=True, right_index=True, how='outer')\n",
    "    output_nonsig_df.to_parquet(output_prefix.joinpath('nonsignal_seq_blast_tblastn.parq'))\n",
    "\n",
    "run_protein_search_tools(input_vtx, output_prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['docker', 'run', '--rm', '-v', '/efs/databases/blast:/blast/blastdb', '-v', '/home/ec2-user/repos/dashboard/cache/protein_data:/blast/data', 'ncbi/blast', 'blastp', '-task', 'blastp-fast', '-outfmt', '6 qaccver saccver stitle bitscore qcovs length pident gaps evalue', '-num_threads', '8', '-db', 'mouse.protein.faa', '-query', '/blast/data/protein_tools_input.fasta', '-out', '/blast/data/protein_tools_input.blastp.out'], returncode=0)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "    header = ['vtx_id', 'blastp_hit_id', 'blastp_description', 'blastp_score',\n",
    "          'blastp_query_coverage', 'blastp_align_length', 'blastp_align_identity', \n",
    "          'blastp_gaps', 'blastp_evalue']\n",
    "    blast_db_path = pathlib.Path('/efs/databases/blast')\n",
    "    blast_db = '-db mouse.protein.faa'\n",
    "    output_fmt = '6 qaccver saccver stitle bitscore qcovs length pident gaps evalue'\n",
    "    options = f'-outfmt \"{output_fmt}\" -num_threads 8'\n",
    "    query = f'-query /blast/data/{input_vtx.name}'\n",
    "    output = f'-out /blast/data/{input_vtx.stem}.blastp.out'\n",
    "    base_cmd = f'docker run --rm -v {blast_db_path}:/blast/blastdb -v {output_prefix}:/blast/data ncbi/blast'\n",
    "    full_cmd = f'{base_cmd} blastp -task blastp-fast {options} {blast_db} {query} {output}'\n",
    "    subprocess.run(shlex.split(full_cmd))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "    blastp_df = pd.read_csv(output_prefix.joinpath(f'{input_vtx.stem}.blastp.out'), sep='\\t', names=header)\n",
    "    bdf = blastp_df.sort_values(by='blastp_score', ascending=False).groupby('vtx_id').first()\n",
    "    bdf.to_parquet(pathlib.Path(output_prefix).joinpath('mouse_blastp.parq'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "blastp_hit_id                                                                      XP_006515041.1\n",
       "blastp_description       XP_006515041.1 phosphatidate phosphatase LPIN1 isoform X3 [Mus musculus]\n",
       "blastp_score                                                                                 25.8\n",
       "blastp_query_coverage                                                                          56\n",
       "blastp_align_length                                                                            23\n",
       "blastp_align_identity                                                                      52.174\n",
       "blastp_gaps                                                                                     0\n",
       "blastp_evalue                                                                                 3.9\n",
       "Name: VTX-0849999, dtype: object"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdf = pd.read_parquet('/home/ec2-user/repos/dashboard/cache/protein_data/mouse_blastp.parq')\n",
    "bdf.loc['VTX-0849999']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'blastp_score'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1206453/1259085455.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mblastp_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_prefix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoinpath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'{input_vtx.stem}.blastp.out'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnames\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mheader\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mbdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblastp_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'blastp_score'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'vtx_id'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfirst\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mbdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda/envs/veliadash/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, by, axis, ascending, inplace, kind, na_position, ignore_index, key)\u001b[0m\n\u001b[1;32m   6754\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6755\u001b[0m             \u001b[0;31m# len(by) == 1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6756\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6757\u001b[0m             \u001b[0mby\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mby\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6758\u001b[0;31m             \u001b[0mk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_label_or_level_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6759\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6760\u001b[0m             \u001b[0;31m# need to rewrap column in Series to apply key function\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6761\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mkey\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda/envs/veliadash/lib/python3.11/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1774\u001b[0m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mxs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mother_axes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1775\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_level_reference\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1776\u001b[0m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_level_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1777\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1778\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1779\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1780\u001b[0m         \u001b[0;31m# Check for duplicates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1781\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'blastp_score'"
     ]
    }
   ],
   "source": [
    "    blastp_df = pd.read_csv(output_prefix.joinpath(f'{input_vtx.stem}.blastp.out'), sep='\\t', names=header)\n",
    "    bdf = blastp_df.sort_values(by='blastp_score', ascending=False).groupby('vtx_id').first()\n",
    "    bdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run blast against sequneces with signal peptide removed\n",
    "#Blast Tools\n",
    "blast_db_path = pathlib.Path('/efs/databases/blast')\n",
    "blast_db = '-db mouse.protein.genbank.faa'\n",
    "output_fmt = '6 qaccver saccver stitle bitscore qcovs length pident gaps evalue'\n",
    "options = f'-outfmt \"{output_fmt}\" -num_threads 8'\n",
    "query = f'-query /blast/data/{VTX_WITHOUT_SIGNAL_SEQ.name}'\n",
    "output = f'-out /blast/data/{VTX_WITHOUT_SIGNAL_SEQ.stem}.blastp.out'\n",
    "base_cmd = f'docker run --rm -v {blast_db_path}:/blast/blastdb -v {output_prefix}:/blast/data ncbi/blast'\n",
    "full_cmd = f'{base_cmd} blastp {options} {blast_db} {query} {output}'\n",
    "# subprocess.run(shlex.split(full_cmd))\n",
    "\n",
    "header = [\n",
    "    'vtx_id', 'blastp_refseq_id', 'nonsig_blastp_hit_id', 'nonsig_blastp_description', 'nonsig_blastp_score',\n",
    "    'nonsig_blastp_align_length', 'nonsig_blastp_align_identity', 'nonsig_blastp_gaps', 'nonsig_blastp_evalue']\n",
    "\n",
    "non_sig_blastp_df = pd.read_csv(output_prefix.joinpath(f'{VTX_WITHOUT_SIGNAL_SEQ.stem}.blastp.out'), sep='\\t', names=header)\n",
    "\n",
    "header = [\n",
    "    'vtx_id', 'tblastn_refseq_id', 'nonsig_tblastn_hit_id', 'nonsig_tblastn_description', 'nonsig_tblastn_score',\n",
    "    'nonsig_tblastn_align_length', 'nonsig_tblastn_align_identity', 'nonsig_tblastn_gaps', 'nonsig_tblastn_evalue']\n",
    "output = f'-out /blast/data/{VTX_WITHOUT_SIGNAL_SEQ.stem}.tblastn.out'\n",
    "blast_db = '-db mouse.rna.fna'\n",
    "base_cmd = f'docker run --rm -v {blast_db_path}:/blast/blastdb -v {output_prefix}:/blast/data ncbi/blast'\n",
    "full_cmd = f'{base_cmd} tblastn {options} {blast_db} {query} {output}'\n",
    "# subprocess.run(shlex.split(full_cmd))\n",
    "tblastn_df = pd.read_csv(output_prefix.joinpath(f'{VTX_WITHOUT_SIGNAL_SEQ.stem}.tblastn.out'), sep='\\t', names=header)\n",
    "tdf = tblastn_df.sort_values(by='nonsig_tblastn_score', ascending=False).groupby('vtx_id').first()\n",
    "bdf = non_sig_blastp_df.sort_values('nonsig_blastp_score', ascending=False).groupby('vtx_id').first()\n",
    "output_nonsig_df = bdf.merge(tdf, left_index=True, right_index=True, how='outer')\n",
    "output_nonsig_df.to_parquet(output_prefix.joinpath('nonsignal_seq_blast_tblastn.parq'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>blastp_refseq_id</th>\n",
       "      <th>nonsig_blastp_hit_id</th>\n",
       "      <th>nonsig_blastp_description</th>\n",
       "      <th>nonsig_blastp_score</th>\n",
       "      <th>nonsig_blastp_align_length</th>\n",
       "      <th>nonsig_blastp_align_identity</th>\n",
       "      <th>nonsig_blastp_gaps</th>\n",
       "      <th>nonsig_blastp_evalue</th>\n",
       "      <th>tblastn_refseq_id</th>\n",
       "      <th>nonsig_tblastn_hit_id</th>\n",
       "      <th>nonsig_tblastn_description</th>\n",
       "      <th>nonsig_tblastn_score</th>\n",
       "      <th>nonsig_tblastn_align_length</th>\n",
       "      <th>nonsig_tblastn_align_identity</th>\n",
       "      <th>nonsig_tblastn_gaps</th>\n",
       "      <th>nonsig_tblastn_evalue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vtx_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VTX-0006555</th>\n",
       "      <td>AAF15261.1</td>\n",
       "      <td>AAF15261.1 Bv8 variant 3 precursor [Mus musculus]</td>\n",
       "      <td>158.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>95.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.280000e-51</td>\n",
       "      <td>NM_015768.3</td>\n",
       "      <td>NM_015768.3 Mus musculus prokineticin 2 (Prok2), transcript variant 1, mRNA</td>\n",
       "      <td>148.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>75.248</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2.310000e-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0006906</th>\n",
       "      <td>NP_001030077.1</td>\n",
       "      <td>NP_001030077.1 sperm-associated antigen 11B isoform 3 precursor [Mus musculus]</td>\n",
       "      <td>47.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>48.077</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.230000e-08</td>\n",
       "      <td>NM_001034905.2</td>\n",
       "      <td>NM_001034905.2 Mus musculus sperm associated antigen 11B (Spag11b), transcript variant 3, mRNA</td>\n",
       "      <td>47.4</td>\n",
       "      <td>93.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>48.077</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.110000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0009065</th>\n",
       "      <td>EDL22895.1</td>\n",
       "      <td>EDL22895.1 sperm acrosome associated 4 [Mus musculus]</td>\n",
       "      <td>136.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>73.034</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.510000e-42</td>\n",
       "      <td>NM_029806.2</td>\n",
       "      <td>NM_029806.2 Mus musculus Ly6/Plaur domain containing 5 (Lypd5), mRNA</td>\n",
       "      <td>30.8</td>\n",
       "      <td>77.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>32.941</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.800000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0009335</th>\n",
       "      <td>AAH31737.1</td>\n",
       "      <td>AAH31737.1 Gene model 75, (NCBI) [Mus musculus]</td>\n",
       "      <td>143.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>53.043</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.350000e-44</td>\n",
       "      <td>NM_203508.2</td>\n",
       "      <td>NM_203508.2 Mus musculus epididymal protein 3B (Eddm3b), mRNA</td>\n",
       "      <td>142.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>53.043</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.990000e-40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0009367</th>\n",
       "      <td>AAI47100.1</td>\n",
       "      <td>AAI47100.1 Gene model 94, (NCBI) [Mus musculus]</td>\n",
       "      <td>84.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>63.636</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.910000e-22</td>\n",
       "      <td>NM_001033280.2</td>\n",
       "      <td>NM_001033280.2 Mus musculus predicted gene 94 (Gm94), mRNA</td>\n",
       "      <td>56.2</td>\n",
       "      <td>62.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>64.103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.090000e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0860429</th>\n",
       "      <td>NP_001074752.1</td>\n",
       "      <td>NP_001074752.1 NALCN channel auxiliary factor 2 [Mus musculus]</td>\n",
       "      <td>30.4</td>\n",
       "      <td>81.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>38.462</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.300000e-01</td>\n",
       "      <td>XM_036160193.1</td>\n",
       "      <td>XM_036160193.1 PREDICTED: Mus musculus UDP-Gal:betaGlcNAc beta 1,3-galactosyltransferase, polypeptide 5 (B3galt5), transcript variant X4, mRNA</td>\n",
       "      <td>28.9</td>\n",
       "      <td>67.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>34.375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.300000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0860430</th>\n",
       "      <td>sp|A2BDP1.1|NALF2_MOUSE</td>\n",
       "      <td>sp|A2BDP1.1|NALF2_MOUSE RecName: Full=NALCN channel auxiliary factor 2; AltName: Full=Protein TED; AltName: Full=Transmembrane protein 28; AltName: Full=Transmembrane protein FAM155B</td>\n",
       "      <td>30.4</td>\n",
       "      <td>81.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>38.462</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.300000e-01</td>\n",
       "      <td>XM_036160191.1</td>\n",
       "      <td>XM_036160191.1 PREDICTED: Mus musculus UDP-Gal:betaGlcNAc beta 1,3-galactosyltransferase, polypeptide 5 (B3galt5), transcript variant X1, mRNA</td>\n",
       "      <td>28.9</td>\n",
       "      <td>67.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>34.375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.300000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0860431</th>\n",
       "      <td>NP_001398496.1</td>\n",
       "      <td>NP_001398496.1 peptidyl-prolyl cis-trans isomerase FKBP14 isoform 3 [Mus musculus]</td>\n",
       "      <td>28.1</td>\n",
       "      <td>66.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>52.632</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.300000e-01</td>\n",
       "      <td>NM_172428.2</td>\n",
       "      <td>NM_172428.2 Mus musculus coiled-coil domain containing 134 (Ccdc134), transcript variant 1, mRNA</td>\n",
       "      <td>25.4</td>\n",
       "      <td>79.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>47.826</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.400000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0860432</th>\n",
       "      <td>BAE43275.1</td>\n",
       "      <td>BAE43275.1 unnamed protein product [Mus musculus]</td>\n",
       "      <td>51.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>45.570</td>\n",
       "      <td>43.0</td>\n",
       "      <td>6.860000e-10</td>\n",
       "      <td>NR_144566.1</td>\n",
       "      <td>NR_144566.1 Mus musculus ubiquitin-conjugating enzyme E2K (Ube2k), transcript variant 4, non-coding RNA</td>\n",
       "      <td>35.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.000000e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0860435</th>\n",
       "      <td>BAE20866.1</td>\n",
       "      <td>BAE20866.1 unnamed protein product [Mus musculus]</td>\n",
       "      <td>84.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>63.636</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.910000e-22</td>\n",
       "      <td>NM_001033280.2</td>\n",
       "      <td>NM_001033280.2 Mus musculus predicted gene 94 (Gm94), mRNA</td>\n",
       "      <td>56.2</td>\n",
       "      <td>62.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>64.103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.090000e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1404 rows  16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    blastp_refseq_id  \\\n",
       "vtx_id                                 \n",
       "VTX-0006555               AAF15261.1   \n",
       "VTX-0006906           NP_001030077.1   \n",
       "VTX-0009065               EDL22895.1   \n",
       "VTX-0009335               AAH31737.1   \n",
       "VTX-0009367               AAI47100.1   \n",
       "...                              ...   \n",
       "VTX-0860429           NP_001074752.1   \n",
       "VTX-0860430  sp|A2BDP1.1|NALF2_MOUSE   \n",
       "VTX-0860431           NP_001398496.1   \n",
       "VTX-0860432               BAE43275.1   \n",
       "VTX-0860435               BAE20866.1   \n",
       "\n",
       "                                                                                                                                                                               nonsig_blastp_hit_id  \\\n",
       "vtx_id                                                                                                                                                                                                \n",
       "VTX-0006555                                                                                                                                       AAF15261.1 Bv8 variant 3 precursor [Mus musculus]   \n",
       "VTX-0006906                                                                                                          NP_001030077.1 sperm-associated antigen 11B isoform 3 precursor [Mus musculus]   \n",
       "VTX-0009065                                                                                                                                   EDL22895.1 sperm acrosome associated 4 [Mus musculus]   \n",
       "VTX-0009335                                                                                                                                         AAH31737.1 Gene model 75, (NCBI) [Mus musculus]   \n",
       "VTX-0009367                                                                                                                                         AAI47100.1 Gene model 94, (NCBI) [Mus musculus]   \n",
       "...                                                                                                                                                                                             ...   \n",
       "VTX-0860429                                                                                                                          NP_001074752.1 NALCN channel auxiliary factor 2 [Mus musculus]   \n",
       "VTX-0860430  sp|A2BDP1.1|NALF2_MOUSE RecName: Full=NALCN channel auxiliary factor 2; AltName: Full=Protein TED; AltName: Full=Transmembrane protein 28; AltName: Full=Transmembrane protein FAM155B   \n",
       "VTX-0860431                                                                                                      NP_001398496.1 peptidyl-prolyl cis-trans isomerase FKBP14 isoform 3 [Mus musculus]   \n",
       "VTX-0860432                                                                                                                                       BAE43275.1 unnamed protein product [Mus musculus]   \n",
       "VTX-0860435                                                                                                                                       BAE20866.1 unnamed protein product [Mus musculus]   \n",
       "\n",
       "             nonsig_blastp_description  nonsig_blastp_score  \\\n",
       "vtx_id                                                        \n",
       "VTX-0006555                      158.0                100.0   \n",
       "VTX-0006906                       47.0                 93.0   \n",
       "VTX-0009065                      136.0                 86.0   \n",
       "VTX-0009335                      143.0                 95.0   \n",
       "VTX-0009367                       84.0                100.0   \n",
       "...                                ...                  ...   \n",
       "VTX-0860429                       30.4                 81.0   \n",
       "VTX-0860430                       30.4                 81.0   \n",
       "VTX-0860431                       28.1                 66.0   \n",
       "VTX-0860432                       51.2                100.0   \n",
       "VTX-0860435                       84.0                100.0   \n",
       "\n",
       "             nonsig_blastp_align_length  nonsig_blastp_align_identity  \\\n",
       "vtx_id                                                                  \n",
       "VTX-0006555                        80.0                        95.000   \n",
       "VTX-0006906                        52.0                        48.077   \n",
       "VTX-0009065                        89.0                        73.034   \n",
       "VTX-0009335                       115.0                        53.043   \n",
       "VTX-0009367                        66.0                        63.636   \n",
       "...                                 ...                           ...   \n",
       "VTX-0860429                        39.0                        38.462   \n",
       "VTX-0860430                        39.0                        38.462   \n",
       "VTX-0860431                        19.0                        52.632   \n",
       "VTX-0860432                        79.0                        45.570   \n",
       "VTX-0860435                        66.0                        63.636   \n",
       "\n",
       "             nonsig_blastp_gaps  nonsig_blastp_evalue tblastn_refseq_id  \\\n",
       "vtx_id                                                                    \n",
       "VTX-0006555                 0.0          2.280000e-51       NM_015768.3   \n",
       "VTX-0006906                 2.0          7.230000e-08    NM_001034905.2   \n",
       "VTX-0009065                 1.0          5.510000e-42       NM_029806.2   \n",
       "VTX-0009335                 1.0          5.350000e-44       NM_203508.2   \n",
       "VTX-0009367                 3.0          2.910000e-22    NM_001033280.2   \n",
       "...                         ...                   ...               ...   \n",
       "VTX-0860429                 0.0          2.300000e-01    XM_036160193.1   \n",
       "VTX-0860430                 0.0          2.300000e-01    XM_036160191.1   \n",
       "VTX-0860431                 0.0          7.300000e-01       NM_172428.2   \n",
       "VTX-0860432                43.0          6.860000e-10       NR_144566.1   \n",
       "VTX-0860435                 3.0          2.910000e-22    NM_001033280.2   \n",
       "\n",
       "                                                                                                                                      nonsig_tblastn_hit_id  \\\n",
       "vtx_id                                                                                                                                                        \n",
       "VTX-0006555                                                                     NM_015768.3 Mus musculus prokineticin 2 (Prok2), transcript variant 1, mRNA   \n",
       "VTX-0006906                                                  NM_001034905.2 Mus musculus sperm associated antigen 11B (Spag11b), transcript variant 3, mRNA   \n",
       "VTX-0009065                                                                            NM_029806.2 Mus musculus Ly6/Plaur domain containing 5 (Lypd5), mRNA   \n",
       "VTX-0009335                                                                                   NM_203508.2 Mus musculus epididymal protein 3B (Eddm3b), mRNA   \n",
       "VTX-0009367                                                                                      NM_001033280.2 Mus musculus predicted gene 94 (Gm94), mRNA   \n",
       "...                                                                                                                                                     ...   \n",
       "VTX-0860429  XM_036160193.1 PREDICTED: Mus musculus UDP-Gal:betaGlcNAc beta 1,3-galactosyltransferase, polypeptide 5 (B3galt5), transcript variant X4, mRNA   \n",
       "VTX-0860430  XM_036160191.1 PREDICTED: Mus musculus UDP-Gal:betaGlcNAc beta 1,3-galactosyltransferase, polypeptide 5 (B3galt5), transcript variant X1, mRNA   \n",
       "VTX-0860431                                                NM_172428.2 Mus musculus coiled-coil domain containing 134 (Ccdc134), transcript variant 1, mRNA   \n",
       "VTX-0860432                                         NR_144566.1 Mus musculus ubiquitin-conjugating enzyme E2K (Ube2k), transcript variant 4, non-coding RNA   \n",
       "VTX-0860435                                                                                      NM_001033280.2 Mus musculus predicted gene 94 (Gm94), mRNA   \n",
       "\n",
       "             nonsig_tblastn_description  nonsig_tblastn_score  \\\n",
       "vtx_id                                                          \n",
       "VTX-0006555                       148.0                 100.0   \n",
       "VTX-0006906                        47.4                  93.0   \n",
       "VTX-0009065                        30.8                  77.0   \n",
       "VTX-0009335                       142.0                  95.0   \n",
       "VTX-0009367                        56.2                  62.0   \n",
       "...                                 ...                   ...   \n",
       "VTX-0860429                        28.9                  67.0   \n",
       "VTX-0860430                        28.9                  67.0   \n",
       "VTX-0860431                        25.4                  79.0   \n",
       "VTX-0860432                        35.0                  47.0   \n",
       "VTX-0860435                        56.2                  62.0   \n",
       "\n",
       "             nonsig_tblastn_align_length  nonsig_tblastn_align_identity  \\\n",
       "vtx_id                                                                    \n",
       "VTX-0006555                        101.0                         75.248   \n",
       "VTX-0006906                         52.0                         48.077   \n",
       "VTX-0009065                         85.0                         32.941   \n",
       "VTX-0009335                        115.0                         53.043   \n",
       "VTX-0009367                         39.0                         64.103   \n",
       "...                                  ...                            ...   \n",
       "VTX-0860429                         32.0                         34.375   \n",
       "VTX-0860430                         32.0                         34.375   \n",
       "VTX-0860431                         23.0                         47.826   \n",
       "VTX-0860432                         17.0                        100.000   \n",
       "VTX-0860435                         39.0                         64.103   \n",
       "\n",
       "             nonsig_tblastn_gaps  nonsig_tblastn_evalue  \n",
       "vtx_id                                                   \n",
       "VTX-0006555                 21.0           2.310000e-42  \n",
       "VTX-0006906                  2.0           4.110000e-07  \n",
       "VTX-0009065                 16.0           1.800000e+00  \n",
       "VTX-0009335                  1.0           1.990000e-40  \n",
       "VTX-0009367                  0.0           3.090000e-10  \n",
       "...                          ...                    ...  \n",
       "VTX-0860429                  0.0           1.300000e+00  \n",
       "VTX-0860430                  0.0           1.300000e+00  \n",
       "VTX-0860431                  0.0           8.400000e+00  \n",
       "VTX-0860432                  0.0           4.000000e-03  \n",
       "VTX-0860435                  0.0           3.090000e-10  \n",
       "\n",
       "[1404 rows x 16 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_nonsig_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/home/ec2-user/repos/dashboard/cache/protein_data/vtx_aa_seq_signal_sequence_removed.tblastn.out'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m tblastn_df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutput_prefix\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoinpath\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mVTX_WITHOUT_SIGNAL_SEQ\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstem\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m.tblastn.out\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnames\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheader\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda/envs/veliadash/lib/python3.11/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    900\u001b[0m     dialect,\n\u001b[1;32m    901\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    909\u001b[0m )\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda/envs/veliadash/lib/python3.11/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/anaconda/envs/veliadash/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda/envs/veliadash/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1659\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1665\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1668\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1670\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1672\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/anaconda/envs/veliadash/lib/python3.11/site-packages/pandas/io/common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    854\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    855\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    858\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 859\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(\n\u001b[1;32m    860\u001b[0m             handle,\n\u001b[1;32m    861\u001b[0m             ioargs\u001b[38;5;241m.\u001b[39mmode,\n\u001b[1;32m    862\u001b[0m             encoding\u001b[38;5;241m=\u001b[39mioargs\u001b[38;5;241m.\u001b[39mencoding,\n\u001b[1;32m    863\u001b[0m             errors\u001b[38;5;241m=\u001b[39merrors,\n\u001b[1;32m    864\u001b[0m             newline\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    865\u001b[0m         )\n\u001b[1;32m    866\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    867\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    868\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/home/ec2-user/repos/dashboard/cache/protein_data/vtx_aa_seq_signal_sequence_removed.tblastn.out'"
     ]
    }
   ],
   "source": [
    "tblastn_df = pd.read_csv(output_prefix.joinpath(f'{VTX_WITHOUT_SIGNAL_SEQ.stem}.tblastn.out'), sep='\\t', names=header)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'docker run --rm -v /efs/databases/blast:/blast/blastdb -v /home/ec2-user/repos/dashboard/cache/protein_data:/blast/data ncbi/blast blastp -outfmt \"6 qaccver saccver stitle bitscore qcovs length pident gaps evalue\" -num_threads 8 -db mouse.protein.genbank.faa -query /blast/data/vtx_aa_seq_signal_sequence_removed.fa -out /blast/data/vtx_aa_seq_signal_sequence_removed.tblastp.out'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GRCh38_latest_protein.faa\n",
      "TMRs.gff3\n",
      "blastp.results.json\n",
      "deepsig.results\n",
      "deeptmhmm_results.md\n",
      "ensembl_isoform.csv\n",
      "ensembl_proteins.fa\n",
      "esmfold.jsonlines\n",
      "isoforms_search.parq\n",
      "kibby.out\n",
      "mouse_blastp.csv\n",
      "predicted_topologies.3line\n",
      "prediction_results.txt\n",
      "processed_entries.fasta\n",
      "protein_tools_input.blastp.out\n",
      "protein_tools_input.fasta\n",
      "protein_tools_input.tblastn.out\n",
      "protein_tools_input_GRCh38_latest_protein_alignments.m8\n",
      "protein_tools_input_ensembl_proteins_alignments.m8\n",
      "protein_tools_input_swissprot_proteins_alignments.m8\n",
      "refseq_isoform.csv\n",
      "region_output.gff3\n",
      "results_summary.signalp5\n",
      "sequence_features_scores.csv\n",
      "sequence_features_strings.csv\n",
      "swissprot_isoform.csv\n",
      "swissprot_proteins.fa\n",
      "tblastn.csv\n",
      "tblastn.results.csv\n",
      "\u001b[0m\u001b[01;34mtmp\u001b[0m/\n",
      "vtx_aa_seq_signal_sequence_removed.blastp.out\n",
      "vtx_aa_seq_signal_sequence_removed.fa\n",
      "xena.db\n"
     ]
    }
   ],
   "source": [
    "ls /home/ec2-user/repos/dashboard/cache/protein_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'docker run --rm -v /efs/databases/blast:/blast/blastdb -v /home/ec2-user/repos/dashboard/cache/protein_data:/blast/data ncbi/blast blastp -outfmt \"6 qaccver saccver stitle bitscore qcovs length pident gaps evalue\" -num_threads 8 -db mouse.protein.genbank.faa -query /blast/data/vtx_aa_seq_signal_sequence_removed.fa -out /blast/data/vtx_aa_seq_signal_sequence_removed.blastp.out'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "    isoform_df = pd.DataFrame([swissprot_isoform_df['swissprot_isoform'], ensembl_isoform_df['ensembl_isoform'], refseq_isoform_df['refseq_isoform']]).T\n",
    "    # isoform_df.replace(pd.NA, ['None'], inplace=True)\n",
    "    isoform_df.to_parquet(pathlib.Path(output_prefix).joinpath('isoforms_search.parq'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from ast import literal_eval\n",
    "\n",
    "from dashboard.data_load import load_sorf_df_conformed, load_protein_feature_string_representations\n",
    "from dashboard.etl import CACHE_DIR, DATA_DIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_sorf_df_conformed()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BLASTp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('isoform_data/nonsignal_seq_aa.fa', 'w') as outfile:\n",
    "    for i, row in df[df['nonsignal_seqs'] != ''].iterrows():\n",
    "        outfile.write(f\">{row.name}\\n{row['nonsignal_seqs']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "vtx_fasta = Path('isoform_data/nonsignal_seq_aa.fa').absolute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = Path('isoform_data').absolute()\n",
    "blast_db_path = Path('/home/ubuntu/velia-data-dev/VDC_004_annotation/primary/blast/')\n",
    "blast_db = '-db mouse.protein.genbank.faa'\n",
    "output_fmt = '6 qaccver saccver stitle bitscore qcovs length pident gaps evalue'\n",
    "options = f'-outfmt \"{output_fmt}\" -num_threads 8'\n",
    "query = f'-query /blast/data/{vtx_fasta.name}'\n",
    "output = f'-out /blast/data/{vtx_fasta.stem}.blastp.out'\n",
    "\n",
    "base_cmd = f'docker run --rm -it -v {blast_db_path}:/blast/blastdb -v {data_path}:/blast/data ncbi/blast'\n",
    "full_cmd = f'{base_cmd} blastp {options} {blast_db} {query} {output}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "!$full_cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = ['vtx_id', 'blastp_hit_id', 'blastp_description', 'blastp_score',\n",
    "          'blastp_query_coverage', 'blastp_align_length', 'blastp_align_identity', \n",
    "          'blastp_gaps', 'blastp_evalue']\n",
    "\n",
    "blastp_df = pd.read_csv(f'isoform_data/{vtx_fasta.stem}.blastp.out', sep='\\t', names=header)\n",
    "bdf = blastp_df.sort_values(by='blastp_score', ascending=False).groupby('vtx_id').first()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## tBLASTn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = Path('isoform_data').absolute()\n",
    "blast_db_path = Path('/home/ubuntu/velia-data-dev/VDC_004_annotation/primary/blast/')\n",
    "blast_db = '-db mouse.rna.fna'\n",
    "output_fmt = '6 qaccver saccver stitle score qcovs length pident gaps evalue'\n",
    "options = f'-outfmt \"{output_fmt}\" -num_threads 8'\n",
    "query = f'-query /blast/data/{vtx_fasta.name}'\n",
    "output = f'-out /blast/data/{vtx_fasta.stem}.tblastn.out'\n",
    "\n",
    "base_cmd = f'docker run --rm -it -v {blast_db_path}:/blast/blastdb -v {data_path}:/blast/data ncbi/blast'\n",
    "full_cmd = f'{base_cmd} tblastn {options} {blast_db} {query} {output}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: [tblastn] Query_434 VTX-0851616: Could not calculate ungapped Karlin-Altschul parameters due to an invalid query sequence or its translation. Please verify the query sequence(s) and/or filtering options \n",
      "Warning: [tblastn] Query_1505 VTX-0850135: Could not calculate ungapped Karlin-Altschul parameters due to an invalid query sequence or its translation. Please verify the query sequence(s) and/or filtering options \n"
     ]
    }
   ],
   "source": [
    "!$full_cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = ['vtx_id', 'tblastn_hit_id', 'tblastn_description', 'tblastn_score',\n",
    "          'tblastn_query_coverage', 'tblastn_align_length', 'tblastn_align_identity', \n",
    "          'tblastn_gaps', 'tblastn_evalue']\n",
    "\n",
    "tblastn_df = pd.read_csv(f'isoform_data/{vtx_fasta.stem}.tblastn.out', sep='\\t', names=header)\n",
    "tdf = tblastn_df.sort_values(by='tblastn_score', ascending=False).groupby('vtx_id').first()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../cache_updates/all_vtx_gencode_231218.txt', 'w') as outfile:\n",
    "    for i, row in df.iterrows():\n",
    "        outfile.write(f\"{row.name}\\n\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "isoform_data_path = Path('/home/ubuntu/dashboard/scripts/notebooks/isoform_data/')\n",
    "vtx_fasta = isoform_data_path.joinpath('nonsignal_seq_aa.fa')\n",
    "\n",
    "header = [\n",
    "    'vtx_id', 'blastp_refseq_id', 'nonsig_blastp_hit_id', 'nonsig_blastp_description', 'blastp_score',\n",
    "    'nonsig_blastp_align_length', 'nonsig_blastp_align_identity', 'blastp_gaps', 'nonsig_blastp_evalue']\n",
    "\n",
    "blastp_df = pd.read_csv(isoform_data_path.joinpath(f'{vtx_fasta.stem}.blastp.out'), sep='\\t', names=header)\n",
    "bdf = blastp_df.sort_values(by='blastp_score', ascending=False).groupby('vtx_id').first()\n",
    "\n",
    "header = [\n",
    "    'vtx_id', 'tblastn_refseq_id', 'nonsig_tblastn_hit_id', 'nonsig_tblastn_description', 'tblastn_score',\n",
    "    'nonsig_tblastn_align_length', 'nonsig_tblastn_align_identity', 'tblastn_gaps', 'nonsig_tblastn_evalue']\n",
    "\n",
    "tblastn_df = pd.read_csv(isoform_data_path.joinpath(f'{vtx_fasta.stem}.tblastn.out'), sep='\\t', names=header)\n",
    "tdf = tblastn_df.sort_values(by='tblastn_score', ascending=False).groupby('vtx_id').first()\n",
    "\n",
    "df = df.merge(bdf, left_index=True, right_index=True, how='left')\n",
    "df = df.merge(tdf, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>show_details</th>\n",
       "      <th>vtx_id</th>\n",
       "      <th>aa_length</th>\n",
       "      <th>ucsc_track</th>\n",
       "      <th>source</th>\n",
       "      <th>screening_phase_id</th>\n",
       "      <th>screening_phase</th>\n",
       "      <th>genscript_id</th>\n",
       "      <th>orf_xrefs</th>\n",
       "      <th>protein_xrefs</th>\n",
       "      <th>...</th>\n",
       "      <th>blastp_gaps</th>\n",
       "      <th>nonsig_blastp_evalue</th>\n",
       "      <th>tblastn_refseq_id</th>\n",
       "      <th>nonsig_tblastn_hit_id</th>\n",
       "      <th>nonsig_tblastn_description</th>\n",
       "      <th>tblastn_score</th>\n",
       "      <th>nonsig_tblastn_align_length</th>\n",
       "      <th>nonsig_tblastn_align_identity</th>\n",
       "      <th>tblastn_gaps</th>\n",
       "      <th>nonsig_tblastn_evalue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vtx_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VTX-0852017</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0852017</td>\n",
       "      <td>54</td>\n",
       "      <td>chr1:-1--1</td>\n",
       "      <td>[velia_phase1_Bona fide]</td>\n",
       "      <td>-1</td>\n",
       "      <td>Phase 1</td>\n",
       "      <td></td>\n",
       "      <td>[NaN, PIGBOS1 NT control ]</td>\n",
       "      <td>[PIGBOS1 NT control ]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0852018</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0852018</td>\n",
       "      <td>53</td>\n",
       "      <td>chr1:-1--1</td>\n",
       "      <td>[velia_phase1_Bona fide]</td>\n",
       "      <td>-1</td>\n",
       "      <td>Phase 1</td>\n",
       "      <td></td>\n",
       "      <td>[HOXB-AS3 NT control , NaN]</td>\n",
       "      <td>[HOXB-AS3 NT control , HOXB-AS3 NT control , N...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0852019</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0852019</td>\n",
       "      <td>45</td>\n",
       "      <td>chr1:-1--1</td>\n",
       "      <td>[velia_phase2_Bonafide_Bianca]</td>\n",
       "      <td>Phase 2_154</td>\n",
       "      <td>Phase 2</td>\n",
       "      <td>U9096HI220-282</td>\n",
       "      <td>[Phase 2_154, U9096HI220-282, ZBTB37]</td>\n",
       "      <td>[Phase 2_154, Phase 4_352, sORF2154, sORF4352,...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0852020</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0852020</td>\n",
       "      <td>42</td>\n",
       "      <td>chr1:-1--1</td>\n",
       "      <td>[velia_phase2_Bianca_Chen]</td>\n",
       "      <td>Phase 2_155</td>\n",
       "      <td>Phase 2</td>\n",
       "      <td>U9096HI220-246</td>\n",
       "      <td>[17-242, 6xhis-HSA-HRV3C-sORF2155_1_02, Phase ...</td>\n",
       "      <td>[17-242, 57729, 6xhis-HSA-HRV3C-sORF2155_1_02,...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0852021</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0852021</td>\n",
       "      <td>59</td>\n",
       "      <td>chr1:-1--1</td>\n",
       "      <td>[velia_phase2_Bianca_Chen]</td>\n",
       "      <td>Phase 2_156</td>\n",
       "      <td>Phase 2</td>\n",
       "      <td>U312HHH170-182</td>\n",
       "      <td>[Phase 2_156, U312HHH170-182, ZFAND2A_1199767_...</td>\n",
       "      <td>[c7riboseqorf5, Phase 2_156, Phase 4_727, sORF...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0851963</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0851963</td>\n",
       "      <td>79</td>\n",
       "      <td>chr1:120005395-120029951</td>\n",
       "      <td>[velia_phase3_nan]</td>\n",
       "      <td>Phase 3_94</td>\n",
       "      <td>Phase 3</td>\n",
       "      <td>U738YHG260-500</td>\n",
       "      <td>[NOTCH2/ENST00000579475.7/mRNA/5'UTR, NOTCH2/N...</td>\n",
       "      <td>[NOTCH2/ENST00000579475.7/mRNA/5'UTR, NOTCH2/N...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0851964</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0851964</td>\n",
       "      <td>81</td>\n",
       "      <td>chr12:6388403-6388942</td>\n",
       "      <td>[velia_phase3_nan]</td>\n",
       "      <td>U738YHG260-508</td>\n",
       "      <td>Phase 3</td>\n",
       "      <td>U738YHG260-508</td>\n",
       "      <td>[LTBR/ENST00000543542.1/ncRNA/-, U738YHG260-508]</td>\n",
       "      <td>[LTBR/ENST00000543542.1/ncRNA/-, NaN, U738YHG2...</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.1</td>\n",
       "      <td>XM_036165897.1</td>\n",
       "      <td>XM_036165897.1 PREDICTED: Mus musculus lymphot...</td>\n",
       "      <td>93.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>73.913</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0851965</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0851965</td>\n",
       "      <td>81</td>\n",
       "      <td>chr4:125487463-125490005</td>\n",
       "      <td>[velia_phase3_nan]</td>\n",
       "      <td>U738YHG260-510</td>\n",
       "      <td>Phase 3</td>\n",
       "      <td>U738YHG260-510</td>\n",
       "      <td>[FAT4/NM_001291285.1/mRNA/CDS, U738YHG260-510]</td>\n",
       "      <td>[FAT4/NM_001291285.1/mRNA/CDS, NaN, U738YHG260...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0851966</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0851966</td>\n",
       "      <td>82</td>\n",
       "      <td>chr6:15246540-15452073</td>\n",
       "      <td>[velia_phase3_nan]</td>\n",
       "      <td>U738YHG260-516</td>\n",
       "      <td>Phase 3</td>\n",
       "      <td>U738YHG260-516</td>\n",
       "      <td>[JARID2/XM_017010834.2/mRNA/-, U738YHG260-516]</td>\n",
       "      <td>[JARID2/XM_017010834.2/mRNA/-, NaN, U738YHG260...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VTX-0851967</th>\n",
       "      <td>False</td>\n",
       "      <td>VTX-0851967</td>\n",
       "      <td>82</td>\n",
       "      <td>chr1:119955155-119963633</td>\n",
       "      <td>[velia_phase3_nan]</td>\n",
       "      <td>Phase 3_97</td>\n",
       "      <td>Phase 3</td>\n",
       "      <td>U738YHG260-518</td>\n",
       "      <td>[NOTCH2/ENST00000479412.2/ncRNA/-, NOTCH2/ENST...</td>\n",
       "      <td>[NOTCH2/ENST00000479412.2/ncRNA/-, NOTCH2/ENST...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12609 rows  73 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             show_details       vtx_id  aa_length                ucsc_track  \\\n",
       "vtx_id                                                                        \n",
       "VTX-0852017         False  VTX-0852017         54                chr1:-1--1   \n",
       "VTX-0852018         False  VTX-0852018         53                chr1:-1--1   \n",
       "VTX-0852019         False  VTX-0852019         45                chr1:-1--1   \n",
       "VTX-0852020         False  VTX-0852020         42                chr1:-1--1   \n",
       "VTX-0852021         False  VTX-0852021         59                chr1:-1--1   \n",
       "...                   ...          ...        ...                       ...   \n",
       "VTX-0851963         False  VTX-0851963         79  chr1:120005395-120029951   \n",
       "VTX-0851964         False  VTX-0851964         81     chr12:6388403-6388942   \n",
       "VTX-0851965         False  VTX-0851965         81  chr4:125487463-125490005   \n",
       "VTX-0851966         False  VTX-0851966         82    chr6:15246540-15452073   \n",
       "VTX-0851967         False  VTX-0851967         82  chr1:119955155-119963633   \n",
       "\n",
       "                                     source screening_phase_id  \\\n",
       "vtx_id                                                           \n",
       "VTX-0852017        [velia_phase1_Bona fide]                 -1   \n",
       "VTX-0852018        [velia_phase1_Bona fide]                 -1   \n",
       "VTX-0852019  [velia_phase2_Bonafide_Bianca]        Phase 2_154   \n",
       "VTX-0852020      [velia_phase2_Bianca_Chen]        Phase 2_155   \n",
       "VTX-0852021      [velia_phase2_Bianca_Chen]        Phase 2_156   \n",
       "...                                     ...                ...   \n",
       "VTX-0851963              [velia_phase3_nan]         Phase 3_94   \n",
       "VTX-0851964              [velia_phase3_nan]     U738YHG260-508   \n",
       "VTX-0851965              [velia_phase3_nan]     U738YHG260-510   \n",
       "VTX-0851966              [velia_phase3_nan]     U738YHG260-516   \n",
       "VTX-0851967              [velia_phase3_nan]         Phase 3_97   \n",
       "\n",
       "            screening_phase    genscript_id  \\\n",
       "vtx_id                                        \n",
       "VTX-0852017         Phase 1                   \n",
       "VTX-0852018         Phase 1                   \n",
       "VTX-0852019         Phase 2  U9096HI220-282   \n",
       "VTX-0852020         Phase 2  U9096HI220-246   \n",
       "VTX-0852021         Phase 2  U312HHH170-182   \n",
       "...                     ...             ...   \n",
       "VTX-0851963         Phase 3  U738YHG260-500   \n",
       "VTX-0851964         Phase 3  U738YHG260-508   \n",
       "VTX-0851965         Phase 3  U738YHG260-510   \n",
       "VTX-0851966         Phase 3  U738YHG260-516   \n",
       "VTX-0851967         Phase 3  U738YHG260-518   \n",
       "\n",
       "                                                     orf_xrefs  \\\n",
       "vtx_id                                                           \n",
       "VTX-0852017                         [NaN, PIGBOS1 NT control ]   \n",
       "VTX-0852018                        [HOXB-AS3 NT control , NaN]   \n",
       "VTX-0852019              [Phase 2_154, U9096HI220-282, ZBTB37]   \n",
       "VTX-0852020  [17-242, 6xhis-HSA-HRV3C-sORF2155_1_02, Phase ...   \n",
       "VTX-0852021  [Phase 2_156, U312HHH170-182, ZFAND2A_1199767_...   \n",
       "...                                                        ...   \n",
       "VTX-0851963  [NOTCH2/ENST00000579475.7/mRNA/5'UTR, NOTCH2/N...   \n",
       "VTX-0851964   [LTBR/ENST00000543542.1/ncRNA/-, U738YHG260-508]   \n",
       "VTX-0851965     [FAT4/NM_001291285.1/mRNA/CDS, U738YHG260-510]   \n",
       "VTX-0851966     [JARID2/XM_017010834.2/mRNA/-, U738YHG260-516]   \n",
       "VTX-0851967  [NOTCH2/ENST00000479412.2/ncRNA/-, NOTCH2/ENST...   \n",
       "\n",
       "                                                 protein_xrefs  ...  \\\n",
       "vtx_id                                                          ...   \n",
       "VTX-0852017                              [PIGBOS1 NT control ]  ...   \n",
       "VTX-0852018  [HOXB-AS3 NT control , HOXB-AS3 NT control , N...  ...   \n",
       "VTX-0852019  [Phase 2_154, Phase 4_352, sORF2154, sORF4352,...  ...   \n",
       "VTX-0852020  [17-242, 57729, 6xhis-HSA-HRV3C-sORF2155_1_02,...  ...   \n",
       "VTX-0852021  [c7riboseqorf5, Phase 2_156, Phase 4_727, sORF...  ...   \n",
       "...                                                        ...  ...   \n",
       "VTX-0851963  [NOTCH2/ENST00000579475.7/mRNA/5'UTR, NOTCH2/N...  ...   \n",
       "VTX-0851964  [LTBR/ENST00000543542.1/ncRNA/-, NaN, U738YHG2...  ...   \n",
       "VTX-0851965  [FAT4/NM_001291285.1/mRNA/CDS, NaN, U738YHG260...  ...   \n",
       "VTX-0851966  [JARID2/XM_017010834.2/mRNA/-, NaN, U738YHG260...  ...   \n",
       "VTX-0851967  [NOTCH2/ENST00000479412.2/ncRNA/-, NOTCH2/ENST...  ...   \n",
       "\n",
       "            blastp_gaps nonsig_blastp_evalue tblastn_refseq_id  \\\n",
       "vtx_id                                                           \n",
       "VTX-0852017         NaN                  NaN               NaN   \n",
       "VTX-0852018         NaN                  NaN               NaN   \n",
       "VTX-0852019         NaN                  NaN               NaN   \n",
       "VTX-0852020         NaN                  NaN               NaN   \n",
       "VTX-0852021         NaN                  NaN               NaN   \n",
       "...                 ...                  ...               ...   \n",
       "VTX-0851963         NaN                  NaN               NaN   \n",
       "VTX-0851964         4.0                  2.1    XM_036165897.1   \n",
       "VTX-0851965         NaN                  NaN               NaN   \n",
       "VTX-0851966         NaN                  NaN               NaN   \n",
       "VTX-0851967         NaN                  NaN               NaN   \n",
       "\n",
       "                                         nonsig_tblastn_hit_id  \\\n",
       "vtx_id                                                           \n",
       "VTX-0852017                                                NaN   \n",
       "VTX-0852018                                                NaN   \n",
       "VTX-0852019                                                NaN   \n",
       "VTX-0852020                                                NaN   \n",
       "VTX-0852021                                                NaN   \n",
       "...                                                        ...   \n",
       "VTX-0851963                                                NaN   \n",
       "VTX-0851964  XM_036165897.1 PREDICTED: Mus musculus lymphot...   \n",
       "VTX-0851965                                                NaN   \n",
       "VTX-0851966                                                NaN   \n",
       "VTX-0851967                                                NaN   \n",
       "\n",
       "            nonsig_tblastn_description tblastn_score  \\\n",
       "vtx_id                                                 \n",
       "VTX-0852017                        NaN           NaN   \n",
       "VTX-0852018                        NaN           NaN   \n",
       "VTX-0852019                        NaN           NaN   \n",
       "VTX-0852020                        NaN           NaN   \n",
       "VTX-0852021                        NaN           NaN   \n",
       "...                                ...           ...   \n",
       "VTX-0851963                        NaN           NaN   \n",
       "VTX-0851964                       93.0          39.0   \n",
       "VTX-0851965                        NaN           NaN   \n",
       "VTX-0851966                        NaN           NaN   \n",
       "VTX-0851967                        NaN           NaN   \n",
       "\n",
       "            nonsig_tblastn_align_length nonsig_tblastn_align_identity  \\\n",
       "vtx_id                                                                  \n",
       "VTX-0852017                         NaN                           NaN   \n",
       "VTX-0852018                         NaN                           NaN   \n",
       "VTX-0852019                         NaN                           NaN   \n",
       "VTX-0852020                         NaN                           NaN   \n",
       "VTX-0852021                         NaN                           NaN   \n",
       "...                                 ...                           ...   \n",
       "VTX-0851963                         NaN                           NaN   \n",
       "VTX-0851964                        23.0                        73.913   \n",
       "VTX-0851965                         NaN                           NaN   \n",
       "VTX-0851966                         NaN                           NaN   \n",
       "VTX-0851967                         NaN                           NaN   \n",
       "\n",
       "            tblastn_gaps nonsig_tblastn_evalue  \n",
       "vtx_id                                          \n",
       "VTX-0852017          NaN                   NaN  \n",
       "VTX-0852018          NaN                   NaN  \n",
       "VTX-0852019          NaN                   NaN  \n",
       "VTX-0852020          NaN                   NaN  \n",
       "VTX-0852021          NaN                   NaN  \n",
       "...                  ...                   ...  \n",
       "VTX-0851963          NaN                   NaN  \n",
       "VTX-0851964          0.0              0.000129  \n",
       "VTX-0851965          NaN                   NaN  \n",
       "VTX-0851966          NaN                   NaN  \n",
       "VTX-0851967          NaN                   NaN  \n",
       "\n",
       "[12609 rows x 73 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
